### PR TITLE
Add field keywords to rewrite middleware

### PR DESCRIPTION
**Pull request by bwaz**
Add field keywords to the rewrite middleware. First handled keyword will be edns0. Without a keyword used the NewSimpleRule Rule will be used. Logic for handling edns0 will come soon. 

We can add "simple" as well to be backwards compatible.

### CODE REVIEW

File: middleware/rewrite/field.go
Comment (miekg) [inline middleware/rewrite/field.go]: Nit: needs closing dot.

File: middleware/rewrite/field.go
Comment (miekg) [inline middleware/rewrite/field.go]: bogus function body because WIP?

File: middleware/rewrite/setup.go
Comment (miekg) [inline middleware/rewrite/setup.go]: Docs

**Review by miekg — COMMENTED** [review]
I agree with the direction we're going here. The only thing I would ask is to be mindful of the language:
1. extensibility  
2. uniformity with the rest of coredns

**Comment by bwaz** [comment]
@miekg  @johnbelamaric  
Init work, edns0 or any other keywords can be added in a new PR
This change makes field a requirement.

File: middleware/rewrite/README.md
Comment (miekg) [inline middleware/rewrite/README.md]: think we should rename qtype to just `type`, we also just use `class`

File: middleware/rewrite/README.md
Comment (miekg) [inline middleware/rewrite/README.md]: This more of a package level documention then something for a readme. Think this readme's should be focused on usage

File: middleware/rewrite/rewrite.go
Comment (miekg) [inline middleware/rewrite/rewrite.go]: init -> `New returns a new rule.`

**Review by miekg — COMMENTED** [review]
lgtm - modulo some nits. I'll update the 005 release wiki

File: middleware/rewrite/README.md
Comment (miekg) [reply middleware/rewrite/README.md]: and type as well?

File: middleware/rewrite/rewrite_test.go
Comment (miekg) [inline middleware/rewrite/rewrite_test.go]: We should try to use `example.*` zone in these tests (yes, there is a of `miek.nl` and `.nl` out there still)

File: middleware/rewrite/README.md
Comment (bwaz) [reply middleware/rewrite/README.md]: I'll change it to class, type, name

File: middleware/rewrite/rewrite_test.go
Comment (bwaz) [reply middleware/rewrite/rewrite_test.go]: I'm having an issue with testing *
```
diff --git a/middleware/rewrite/rewrite_test.go b/middleware/rewrite/rewrite_test.go
index ffb1b4a..63d2f84 100644
--- a/middleware/rewrite/rewrite_test.go
+++ b/middleware/rewrite/rewrite_test.go
@@ -20,9 +20,10 @@ func TestRewrite(t *testing.T) {
        rw := Rewrite{
                Next: middleware.HandlerFunc(msgPrinter),
                Rules: []Rule{
-                       Fields["qname"].New("from.nl.", "to.nl."),
+                       Fields["name"].New("from.nl.", "to.nl."),
+                       Fields["name"].New("myexample.*", "myexample.com."),
                        Fields["class"].New("CH", "IN"),
-                       Fields["qtype"].New("ANY", "HINFO"),
+                       Fields["type"].New("ANY", "HINFO"),
                },
                noRevert: true,
        }
@@ -35,6 +36,7 @@ func TestRewrite(t *testing.T) {
                toT   uint16
                toC   uint16
        }{
+               {"myexample.org.", dns.TypeA, dns.ClassINET, "myexample.com.", dns.TypeA, dns.ClassINET},
                {"from.nl.", dns.TypeA, dns.ClassINET, "to.nl.", dns.TypeA, dns.ClassINET},
                {"a.nl.", dns.TypeA, dns.ClassINET, "a.nl.", dns.TypeA, dns.ClassINET},
                {"a.nl.", dns.TypeA, dns.ClassCHAOS, "a.nl.", dns.TypeA, dns.ClassINET},

```

```
--- FAIL: TestRewrite (0.00s)
	rewrite_test.go:61: Test 0: Expected Name to be 'myexample.com.' but was 'myexample.org.'

```

Comment (bwaz) [reply middleware/rewrite/rewrite_test.go]: Anything I'm doing wrong with the setup? Can we defer to new issue if functionality is not implemented yet?

File: middleware/rewrite/README.md
Comment (johnbelamaric) [inline middleware/rewrite/README.md]: s/rewriten/rewritten/

File: middleware/rewrite/README.md
Comment (johnbelamaric) [inline middleware/rewrite/README.md:27]: Is this true? Is this what we want? If you can specify multiple re-write middleware, then I assume they all act independently.

File: middleware/rewrite/setup.go
Comment (johnbelamaric) [inline middleware/rewrite/setup.go]: else return an error so the user knows they made a typo...

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Looking good.

File: middleware/rewrite/README.md
Comment (miekg) [reply middleware/rewrite/README.md:27]: It can be done, but it fairly useless, because clients reject responses that are fiddled with, also DNSSEC...

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add field keywo..." ]
>bwaz commented on this pull request.
>
>
>
>> @@ -20,9 +20,9 @@ func TestRewrite(t *testing.T) {
> 	rw := Rewrite{
> 		Next: middleware.HandlerFunc(msgPrinter),
> 		Rules: []Rule{
>-			NewSimpleRule("from.nl.", "to.nl."),
>-			NewSimpleRule("CH", "IN"),
>-			NewSimpleRule("ANY", "HINFO"),
>+			Fields["qname"].New("from.nl.", "to.nl."),
>
>I'm having an issue with testing *
>```
>diff --git a/middleware/rewrite/rewrite_test.go b/middleware/rewrite/rewrite_test.go
>index ffb1b4a..63d2f84 100644
>--- a/middleware/rewrite/rewrite_test.go
>+++ b/middleware/rewrite/rewrite_test.go
>@@ -20,9 +20,10 @@ func TestRewrite(t *testing.T) {
>        rw := Rewrite{
>                Next: middleware.HandlerFunc(msgPrinter),
>                Rules: []Rule{
>-                       Fields["qname"].New("from.nl.", "to.nl."),
>+                       Fields["name"].New("from.nl.", "to.nl."),
>+                       Fields["name"].New("myexample.*", "myexample.com."),
>                        Fields["class"].New("CH", "IN"),
>-                       Fields["qtype"].New("ANY", "HINFO"),
>+                       Fields["type"].New("ANY", "HINFO"),
>                },
>                noRevert: true,
>        }
>@@ -35,6 +36,7 @@ func TestRewrite(t *testing.T) {
>                toT   uint16
>                toC   uint16
>        }{
>+               {"myexample.org.", dns.TypeA, dns.ClassINET, "myexample.com.", dns.TypeA, dns.ClassINET},
>                {"from.nl.", dns.TypeA, dns.ClassINET, "to.nl.", dns.TypeA, dns.ClassINET},
>                {"a.nl.", dns.TypeA, dns.ClassINET, "a.nl.", dns.TypeA, dns.ClassINET},
>                {"a.nl.", dns.TypeA, dns.ClassCHAOS, "a.nl.", dns.TypeA, dns.ClassINET},
>
>```
>
>```
>--- FAIL: TestRewrite (0.00s)
>	rewrite_test.go:61: Test 0: Expected Name to be 'myexample.com.' but was 'myexample.org.'
>
>```

Oh, yeah, this might not work, because you're doing multiple rules chained after
each other (unless you fixed that as well). At least in CoreDNS <= 004 this
chaining never worked.

**Comment by bwaz** [comment]
I suggest splitting any additional work from this point to new PR

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Add field keywo..." ]
>I suggest splitting any additional work from this point to new PR

ack.

**Review by johnbelamaric — APPROVED** [review]

**Comment by johnbelamaric** [comment]
lgtm

**Pull request merged**
